src/System.Text.Json/src/System/Text/Json/Serialization/Converters/DefaultImmutableConverter.cs

```diff
@@ -0,0 +1,142 @@
+
+        private static ConcurrentDictionary<string, object> _createRangeDelegates = new ConcurrentDictionary<string, object>();
+
+        private static readonly ConcurrentDictionary<Type, string> _createRangeDelegateKeys = new ConcurrentDictionary<Type, string>();
+
+        public string GetConstructingTypeName(string immutableCollectionTypeName)
+        {
+            switch (immutableCollectionTypeName)
+            {
+                case ImmutableListGenericTypeName:
+                case ImmutableListGenericInterfaceTypeName:
+                    return ImmutableListTypeName;
+                case ImmutableStackGenericTypeName:
+                case ImmutableStackGenericInterfaceTypeName:
+                    return  ImmutableStackTypeName;
+                case ImmutableQueueGenericTypeName:
+                case ImmutableQueueGenericInterfaceTypeName:
+                    return  ImmutableQueueTypeName;
+                case ImmutableSortedSetGenericTypeName:
+                    return  ImmutableSortedSetTypeName;
+                case ImmutableHashSetGenericTypeName:
+                case ImmutableSetGenericInterfaceTypeName:
+                    return  ImmutableHashSetTypeName;
+                default:
+                    Debug.Fail("Unknown immutable: " + immutableCollectionTypeName);
+                    return null;
+            }
+        }
+
+        // This method creates a TElement[] and populates it with the items in the sourceList argument,
+        // then uses the createRangeDelegateKey argument to identify the appropriate cached
+        // CreateRange<TElement> method to create and return the desired immutable collection type.
+        // The method is constructed for each element type with reflection and cached.
+        public static IEnumerable CreateImmutableCollectionFromList<TElement>(string createRangeDelegateKey, IList sourceList)
+        {
+            Array array = CreateArrayFromList(typeof(TElement), sourceList);
+
+            Debug.Assert(_createRangeDelegates.ContainsKey(createRangeDelegateKey));
+
+            IEnumerable immutableCollection = (IEnumerable)((ImmutableCreateRangeDelegate<TElement>)_createRangeDelegates[createRangeDelegateKey]).DynamicInvoke(array);
+            return immutableCollection;
+        }
+
+        public void RegisterImmutableCollectionType(Type immutableCollectionType, Type elementType, JsonSerializerOptions options)
+        {
+            // If we have registered this immutable collection type:
+            if (_createRangeDelegateKeys.ContainsKey(immutableCollectionType))
+            {
+                // We must have the appropriate ImmutableX.CreateRange<elementType> delegate cached.
+                Debug.Assert(_createRangeDelegates.ContainsKey(_createRangeDelegateKeys[immutableCollectionType]));
+                return;
```

Comment:
Where do we throw a "notsupported" or "invalidoperation" exception for unknown collection types? I assume this Debug.Assert does not get hit for unknown collection type scenarios.

Reply:
I throw a not-supported exception for unsupported immutable collections. I'll refactor following the exception changes we discussed offline.

Reply:
Do we have tests that capture notsupportedexception cases, like `StringCollection` I mentioned earlier?